File: transcription_service/tests/unit/shared/utils/audio_frame_protocol/audio_frame_protocol_test.py
"""
Unit tests for the ScribeAR Audio Frame Protocol (SAFP) codec.
"""

import struct
import zlib

import pytest

from src.shared.utils.audio_frame_protocol import (
    SAFP_VERSION,
    AudioFrameError,
    decode_audio_frame,
    encode_audio_frame,
)
from src.shared.utils.audio_frame_protocol.audio_frame_protocol import (
    FIELD_KEY_CHUNK_ID,
    FIELD_KEY_STAGE_DEPTH,
    SAFP_MAGIC_0,
    SAFP_MAGIC_1,
    WIRE_BYTES,
    WIRE_UINT,
    WIRE_UTF8,
)

AUDIO = bytes([0, 1, 2, 3, 250, 251, 252, 253])


def test_crc_known_answer_matches_typescript():
    """zlib.crc32 must produce the canonical CRC-32 check value, which is the
    same number the TypeScript table-based crc32 produces."""
    assert zlib.crc32(b"123456789") & 0xFFFFFFFF == 0xCBF43926


def test_round_trip_chunk_id_and_sent_at():
    """A frame with both metadata fields decodes back to the same values."""
    frame = encode_audio_frame("id-123", AUDIO, sent_at=1_700_000_000_123.0)
    decoded = decode_audio_frame(frame)

    assert decoded.version == SAFP_VERSION
    assert decoded.chunk_id == "id-123"
    assert decoded.sent_at == 1_700_000_000_123.0
    assert decoded.audio == AUDIO
    assert decoded.unknown_fields == []


def test_round_trip_chunk_id_only():
    """The node -> python hop carries chunk_id only (no sent_at)."""
    frame = encode_audio_frame("abc", AUDIO)
    decoded = decode_audio_frame(frame)

    assert decoded.chunk_id == "abc"
    assert decoded.sent_at is None
    assert decoded.audio == AUDIO


def test_unicode_chunk_id():
    """UTF-8 chunk ids round-trip byte-for-byte."""
    decoded = decode_audio_frame(encode_audio_frame("café-⛄-😀", AUDIO))
    assert decoded.chunk_id == "café-⛄-😀"


def test_empty_audio():
    """A frame may carry no audio payload."""
    decoded = decode_audio_frame(encode_audio_frame("x", b""))
    assert decoded.chunk_id == "x"
    assert decoded.audio == b""


def test_cross_language_frame_shape():
    """Encoded magic + version bytes match the documented envelope, so a
    frame from the TypeScript encoder decodes here and vice versa."""
    frame = encode_audio_frame("x", AUDIO)
    assert frame[0] == SAFP_MAGIC_0
    assert frame[1] == SAFP_MAGIC_1
    assert frame[2] == SAFP_VERSION


def test_forward_compatibility_skips_unknown_field():
    """An unknown field inserted by a newer sender is skipped, and the known
    field + audio are still recovered."""
    chunk_bytes = b"fwd"
    unknown_value = bytes([9, 9, 9])
    fields = bytearray()
    fields += struct.pack(
        "<BBH", FIELD_KEY_CHUNK_ID, WIRE_UTF8, len(chunk_bytes)
    )
    fields += chunk_bytes
    fields += struct.pack("<BBH", 99, WIRE_BYTES, len(unknown_value))
    fields += unknown_value

    body = bytearray()
    body += struct.pack("<BBBB", SAFP_MAGIC_0, SAFP_MAGIC_1, SAFP_VERSION, 2)
    body += fields
    body += AUDIO
    body += struct.pack("<I", zlib.crc32(bytes(body)) & 0xFFFFFFFF)

    decoded = decode_audio_frame(bytes(body))
    assert decoded.chunk_id == "fwd"
    assert decoded.audio == AUDIO
    assert len(decoded.unknown_fields) == 1
    assert decoded.unknown_fields[0][0] == 99


def test_round_trip_stage_depth():
    """An audio-telemetry depth declared by the sender survives the round trip,
    so the transcription service can continue that graph instead of restarting
    its numbering at its own ingress stage."""
    # Arrange / Act
    decoded = decode_audio_frame(
        encode_audio_frame("depth", AUDIO, stage_depth=2)
    )

    # Assert
    assert decoded.stage_depth == 2
    assert decoded.chunk_id == "depth"
    assert decoded.audio == AUDIO
    assert decoded.unknown_fields == []


def test_stage_depth_is_absent_when_the_sender_does_not_meter():
    """No producer writes this field yet (node-server is the intended one), so
    the overwhelmingly common frame carries no depth - and None has to mean
    "the sender does not meter" rather than a depth of zero."""
    # Arrange / Act
    decoded = decode_audio_frame(encode_audio_frame("no-depth", AUDIO))

    # Assert
    assert decoded.stage_depth is None


def test_stage_depth_decodes_from_a_wider_uint():
    """WIRE_UINT is self-describing by length, so a sender that widened the
    field must still be understood - pinning this reader to one byte would make
    such a frame look like it carried an unknown field instead."""
    # Arrange - the same value as a little-endian uint16.
    fields = bytearray()
    fields += struct.pack("<BBH", FIELD_KEY_STAGE_DEPTH, WIRE_UINT, 2)
    fields += struct.pack("<H", 7)

    body = bytearray()
    body += struct.pack("<BBBB", SAFP_MAGIC_0, SAFP_MAGIC_1, SAFP_VERSION, 1)
    body += fields
    body += AUDIO
    body += struct.pack("<I", zlib.crc32(bytes(body)) & 0xFFFFFFFF)

    # Act
    decoded = decode_audio_frame(bytes(body))

    # Assert
    assert decoded.stage_depth == 7
    assert decoded.unknown_fields == []


def test_rejects_a_stage_depth_too_large_for_the_wire_field():
    """The encoder writes one byte, so a caller passing something that does not
    fit has to be told - silently truncating would publish a graph whose depths
    are wrong rather than obviously missing."""
    # Arrange / Act / Assert
    with pytest.raises(AudioFrameError, match="stage_depth"):
        encode_audio_frame("x", AUDIO, stage_depth=256)


def test_rejects_short_buffer():
    """Buffers shorter than the minimum envelope are rejected."""
    with pytest.raises(AudioFrameError):
        decode_audio_frame(b"\x00\x00\x00")


def test_rejects_bad_magic():
    """A frame with the wrong magic is rejected."""
    frame = bytearray(encode_audio_frame("x", AUDIO))
    frame[0] = 0x00
    frame[-4:] = struct.pack("<I", zlib.crc32(bytes(frame[:-4])) & 0xFFFFFFFF)
    with pytest.raises(AudioFrameError, match="magic"):
        decode_audio_frame(bytes(frame))


def test_rejects_crc_mismatch():
    """A single flipped audio byte is caught by the CRC."""
    frame = bytearray(encode_audio_frame("x", AUDIO))
    frame[-5] ^= 0xFF
    with pytest.raises(AudioFrameError, match="CRC"):
        decode_audio_frame(bytes(frame))


def test_rejects_unsupported_version():
    """An unexpected version byte is rejected."""
    frame = bytearray(encode_audio_frame("x", AUDIO))
    frame[2] = 2
    frame[-4:] = struct.pack("<I", zlib.crc32(bytes(frame[:-4])) & 0xFFFFFFFF)
    with pytest.raises(AudioFrameError, match="version"):
        decode_audio_frame(bytes(frame))

File: libs/audio-frame-protocol/tests/unit/audio-frame-protocol.test.ts
import { describe, expect, it } from 'vitest';

import {
  AudioFrameError,
  SAFP_MAGIC_0,
  SAFP_MAGIC_1,
  SAFP_VERSION,
  SafpFieldKey,
  SafpWireType,
  crc32,
  decodeAudioFrame,
  encodeAudioFrame,
} from '#src/index.js';

const audio = new Uint8Array([0, 1, 2, 3, 250, 251, 252, 253]);

describe('crc32', () => {
  it('matches the IEEE known-answer for "123456789"', () => {
    // 0xCBF43926 is the canonical CRC-32 check value; Python's zlib.crc32
    // produces the same number, guaranteeing cross-language agreement.
    const bytes = new TextEncoder().encode('123456789');
    expect(crc32(bytes)).toBe(0xcbf43926);
  });

  it('is empty-safe', () => {
    expect(crc32(new Uint8Array(0))).toBe(0);
  });
});

describe('encodeAudioFrame / decodeAudioFrame', () => {
  it('round-trips chunkId + sentAt + audio', () => {
    const chunkId = crypto.randomUUID();
    const sentAt = 1_700_000_000_123;
    const frame = encodeAudioFrame({ chunkId, sentAt }, audio);
    const decoded = decodeAudioFrame(frame);

    expect(decoded.version).toBe(SAFP_VERSION);
    expect(decoded.chunkId).toBe(chunkId);
    expect(decoded.sentAt).toBe(sentAt);
    expect([...decoded.audio]).toEqual([...audio]);
    expect(decoded.unknownFields).toEqual([]);
  });

  it('round-trips a frame with no sentAt (node -> python hop)', () => {
    const chunkId = 'abc-123';
    const frame = encodeAudioFrame({ chunkId }, audio);
    const decoded = decodeAudioFrame(frame);

    expect(decoded.chunkId).toBe(chunkId);
    expect(decoded.sentAt).toBeNull();
    expect([...decoded.audio]).toEqual([...audio]);
  });

  it('round-trips stageDepth', () => {
    const chunkId = 'depth-123';
    const frame = encodeAudioFrame({ chunkId, stageDepth: 2 }, audio);
    const decoded = decodeAudioFrame(frame);

    expect(decoded.stageDepth).toBe(2);
  });

  it('decodes a frame with no stageDepth as null, not 0, since 0 is a legal depth', () => {
    const chunkId = 'no-depth';
    const frame = encodeAudioFrame({ chunkId }, audio);
    const decoded = decodeAudioFrame(frame);

    expect(decoded.stageDepth).toBeNull();
  });

  it('round-trips stageDepth 0, distinguishing it from absent', () => {
    const chunkId = 'zero-depth';
    const frame = encodeAudioFrame({ chunkId, stageDepth: 0 }, audio);
    const decoded = decodeAudioFrame(frame);

    expect(decoded.stageDepth).toBe(0);
  });

  it('rejects a stageDepth outside the single-byte range, matching the Python encoder', () => {
    expect(() =>
      encodeAudioFrame({ chunkId: 'x', stageDepth: 256 }, audio),
    ).toThrow(AudioFrameError);
    expect(() =>
      encodeAudioFrame({ chunkId: 'x', stageDepth: -1 }, audio),
    ).toThrow(AudioFrameError);
  });

  it('writes the documented magic and version', () => {
    const frame = encodeAudioFrame({ chunkId: 'x' }, audio);
    expect(frame[0]).toBe(SAFP_MAGIC_0);
    expect(frame[1]).toBe(SAFP_MAGIC_1);
    expect(frame[2]).toBe(SAFP_VERSION);
  });

  it('handles an empty audio payload', () => {
    const frame = encodeAudioFrame({ chunkId: 'x' }, new Uint8Array(0));
    const decoded = decodeAudioFrame(frame);
    expect(decoded.chunkId).toBe('x');
    expect(decoded.audio.length).toBe(0);
  });

  it('accepts an ArrayBuffer as input', () => {
    const frame = encodeAudioFrame({ chunkId: 'y' }, audio);
    const copy = frame.slice();
    const decoded = decodeAudioFrame(copy.buffer);
    expect(decoded.chunkId).toBe('y');
  });

  it('preserves unicode chunk ids', () => {
    const chunkId = 'café-⛄-😀';
    const decoded = decodeAudioFrame(encodeAudioFrame({ chunkId }, audio));
    expect(decoded.chunkId).toBe(chunkId);
  });
});

describe('forward compatibility', () => {
  it('skips an unknown field and still finds audio + known fields', () => {
    // Hand-build a frame with an extra unknown field (key 99) between the
    // known fields and the audio, mimicking a newer sender.
    const chunkId = 'fwd';
    const chunkBytes = new TextEncoder().encode(chunkId);
    const unknownValue = new Uint8Array([9, 9, 9]);
    const fields = new Uint8Array(
      4 + chunkBytes.length + (4 + unknownValue.length),
    );
    const fView = new DataView(fields.buffer);
    // field 1: chunkId
    fields[0] = SafpFieldKey.CHUNK_ID;
    fields[1] = SafpWireType.UTF8;
    fView.setUint16(2, chunkBytes.length, true);
    fields.set(chunkBytes, 4);
    // field 2: unknown key 99
    const o = 4 + chunkBytes.length;
    fields[o] = 99;
    fields[o + 1] = SafpWireType.BYTES;
    fView.setUint16(o + 2, unknownValue.length, true);
    fields.set(unknownValue, o + 4);

    const body = new Uint8Array(4 + fields.length + audio.length);
    body[0] = SAFP_MAGIC_0;
    body[1] = SAFP_MAGIC_1;
    body[2] = SAFP_VERSION;
    body[3] = 2; // two fields
    body.set(fields, 4);
    body.set(audio, 4 + fields.length);

    const frame = new Uint8Array(body.length + 4);
    frame.set(body, 0);
    new DataView(frame.buffer).setUint32(body.length, crc32(body), true);

    const decoded = decodeAudioFrame(frame);
    expect(decoded.chunkId).toBe(chunkId);
    expect([...decoded.audio]).toEqual([...audio]);
    expect(decoded.unknownFields).toHaveLength(1);
    expect(decoded.unknownFields[0]?.key).toBe(99);
  });

  it('still finds a known stageDepth alongside a field key it does not recognise', () => {
    // A future sender may add its own field ahead of stageDepth landing on
    // every peer; this proves the two co-exist rather than one masking the
    // other.
    const chunkId = 'mixed';
    const chunkBytes = new TextEncoder().encode(chunkId);
    const unknownValue = new Uint8Array([7, 7]);
    const stageDepthValue = new Uint8Array([4]);

    const fields = new Uint8Array(
      4 +
        chunkBytes.length +
        (4 + unknownValue.length) +
        (4 + stageDepthValue.length),
    );
    const fView = new DataView(fields.buffer);
    fields[0] = SafpFieldKey.CHUNK_ID;
    fields[1] = SafpWireType.UTF8;
    fView.setUint16(2, chunkBytes.length, true);
    fields.set(chunkBytes, 4);

    let o = 4 + chunkBytes.length;
    fields[o] = 200; // unrecognised key
    fields[o + 1] = SafpWireType.BYTES;
    fView.setUint16(o + 2, unknownValue.length, true);
    fields.set(unknownValue, o + 4);

    o += 4 + unknownValue.length;
    fields[o] = SafpFieldKey.STAGE_DEPTH;
    fields[o + 1] = SafpWireType.UINT;
    fView.setUint16(o + 2, stageDepthValue.length, true);
    fields.set(stageDepthValue, o + 4);

    const body = new Uint8Array(4 + fields.length + audio.length);
    body[0] = SAFP_MAGIC_0;
    body[1] = SAFP_MAGIC_1;
    body[2] = SAFP_VERSION;
    body[3] = 3; // three fields
    body.set(fields, 4);
    body.set(audio, 4 + fields.length);

    const frame = new Uint8Array(body.length + 4);
    frame.set(body, 0);
    new DataView(frame.buffer).setUint32(body.length, crc32(body), true);

    const decoded = decodeAudioFrame(frame);
    expect(decoded.chunkId).toBe(chunkId);
    expect(decoded.stageDepth).toBe(4);
    expect(decoded.unknownFields).toHaveLength(1);
    expect(decoded.unknownFields[0]?.key).toBe(200);
    expect([...decoded.audio]).toEqual([...audio]);
  });

  it('an old decoder that never learned about stageDepth still finds chunkId, sentAt, and audio in a frame that carries it', () => {
    // Stand-in for the pre-this-change TypeScript decoder: identical framing
    // loop, but only CHUNK_ID and SENT_AT are special-cased, so stageDepth's
    // bytes fall through to "unknown" exactly like any other field an old
    // reader has never heard of. This is the property that lets node-server
    // start sending stageDepth before every reader has been upgraded.
    function decodeWithoutStageDepthSupport(frame: Uint8Array): {
      chunkId: string | null;
      sentAt: number | null;
      audio: Uint8Array;
      unknownKeys: number[];
    } {
      const view = new DataView(
        frame.buffer,
        frame.byteOffset,
        frame.byteLength,
      );
      const crcOffset = frame.length - 4;
      const fieldCount = frame[3] ?? 0;
      let offset = 4;
      let chunkId: string | null = null;
      let sentAt: number | null = null;
      const unknownKeys: number[] = [];

      for (let i = 0; i < fieldCount; i++) {
        const key = frame[offset] ?? 0;
        const length = view.getUint16(offset + 2, true);
        const valueStart = offset + 4;
        const valueEnd = valueStart + length;
        const value = frame.subarray(valueStart, valueEnd);

        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        if (key === (SafpFieldKey.CHUNK_ID as number)) {
          chunkId = new TextDecoder('utf-8').decode(value);
          // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        } else if (key === (SafpFieldKey.SENT_AT as number) && length === 8) {
          sentAt = new DataView(
            value.buffer,
            value.byteOffset,
            value.byteLength,
          ).getFloat64(0, true);
        } else {
          unknownKeys.push(key);
        }

        offset = valueEnd;
      }

      return {
        chunkId,
        sentAt,
        audio: frame.subarray(offset, crcOffset),
        unknownKeys,
      };
    }

    const chunkId = 'old-reader';
    const sentAt = 42;
    const frame = encodeAudioFrame({ chunkId, sentAt, stageDepth: 5 }, audio);
    const result = decodeWithoutStageDepthSupport(frame);

    expect(result.chunkId).toBe(chunkId);
    expect(result.sentAt).toBe(sentAt);
    expect([...result.audio]).toEqual([...audio]);
    expect(result.unknownKeys).toEqual([SafpFieldKey.STAGE_DEPTH]);
  });
});

describe('validation', () => {
  it('rejects a short buffer', () => {
    expect(() => decodeAudioFrame(new Uint8Array(3))).toThrow(AudioFrameError);
  });

  it('rejects a bad magic', () => {
    const frame = encodeAudioFrame({ chunkId: 'x' }, audio);
    frame[0] = 0x00;
    // recompute CRC so we isolate the magic check
    new DataView(frame.buffer).setUint32(
      frame.length - 4,
      crc32(frame.subarray(0, frame.length - 4)),
      true,
    );
    expect(() => decodeAudioFrame(frame)).toThrow(/magic/i);
  });

  it('rejects a corrupted payload via CRC', () => {
    const frame = encodeAudioFrame({ chunkId: 'x' }, audio);
    const i = frame.length - 5;
    frame[i] = (frame[i] ?? 0) ^ 0xff; // flip a byte of audio
    expect(() => decodeAudioFrame(frame)).toThrow(/CRC/i);
  });

  it('rejects an unsupported version', () => {
    const frame = encodeAudioFrame({ chunkId: 'x' }, audio);
    frame[2] = 2;
    new DataView(frame.buffer).setUint32(
      frame.length - 4,
      crc32(frame.subarray(0, frame.length - 4)),
      true,
    );
    expect(() => decodeAudioFrame(frame)).toThrow(/version/i);
  });

  it('still validates CRC-32 once stageDepth is one of the fields covered by it', () => {
    const frame = encodeAudioFrame({ chunkId: 'x', stageDepth: 9 }, audio);
    // decodeAudioFrame throws on CRC mismatch, so a clean decode here is
    // proof the checksum was computed over the frame with the new field
    // included, not stale from before it landed.
    expect(() => decodeAudioFrame(frame)).not.toThrow();

    const i = frame.length - 5;
    frame[i] = (frame[i] ?? 0) ^ 0xff;
    expect(() => decodeAudioFrame(frame)).toThrow(/CRC/i);
  });
});

describe('stageDepth field key and wire type', () => {
  // A drift guard: if the Python and TypeScript implementations ever
  // disagree on the field key number or wire type for stageDepth, a frame
  // encoded by one and decoded by the other silently mis-parses instead of
  // failing loudly, so pin both constants here rather than trusting the two
  // files to stay in sync by inspection.
  it('keys stageDepth as field 3', () => {
    expect(SafpFieldKey.STAGE_DEPTH).toBe(3);
  });

  it('wire-types stageDepth as WIRE_UINT (0x01)', () => {
    expect(SafpWireType.UINT).toBe(0x01);
  });
});
